jetty-websocket/javax-websocket-tests/src/test/java/org/eclipse/jetty/websocket/javax/tests/server/WebAppClassLoaderTest.java

```diff
@@ -0,0 +1,133 @@
+        private final static CompletableFuture<ClassLoader> onErrorClassLoader = new CompletableFuture<>();
+        private final static CompletableFuture<ClassLoader> onCloseClassLoader = new CompletableFuture<>();
+
+        public MySocket()
+        {
+            constructorClassLoader.complete(Thread.currentThread().getContextClassLoader());
+        }
+
+        @OnOpen
+        public void onOpen(Session session)
+        {
+            onOpenClassLoader.complete(Thread.currentThread().getContextClassLoader());
+        }
+
+        @OnMessage
+        public void onMessage(Session session, String msg)
+        {
+            onMessageClassLoader.complete(Thread.currentThread().getContextClassLoader());
+        }
+
+        @OnError
+        public void onError(Throwable error)
+        {
+            onErrorClassLoader.complete(Thread.currentThread().getContextClassLoader());
+        }
+
+        @OnClose
+        public void onClose(CloseReason closeReason)
+        {
+            onCloseClassLoader.complete(Thread.currentThread().getContextClassLoader());
+        }
+    }
+
+    private static WSServer server;
+    private static WebAppContext webapp;
+
+    @BeforeAll
+    public static void startServer() throws Exception
+    {
+        Path testdir = MavenTestingUtils.getTargetTestingPath(WebAppClassLoaderTest.class.getName());
+        server = new WSServer(testdir, "app");
+        server.createWebInf();
+        server.copyEndpoint(MySocket.class);
+        server.start();
+        webapp = server.createWebAppContext();
+        server.deployWebapp(webapp);
+    }
+
+    @AfterAll
+    public static void stopServer() throws Exception
+    {
+        server.stop();
+    }
+
+    private ClassLoader getClassLoader(String event) throws Exception
+    {
+        ClassLoader webAppClassLoader = webapp.getClassLoader();
+        Class<?> aClass = webAppClassLoader.loadClass(MySocket.class.getName());
+        Field field = aClass.getDeclaredField(event + "ClassLoader");
+        field.setAccessible(true);
+        return ((CompletableFuture<ClassLoader>)field.get(null)).get(5, TimeUnit.SECONDS);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"constructor", "onOpen", "onMessage", "onError", "onClose"})
+    public void test(String event) throws Exception
+    {
+        WebSocketContainer client = ContainerProvider.getWebSocketContainer();
+        EventSocket clientSocket = new EventSocket();
+        Session session = client.connectToServer(clientSocket, server.getWsUri().resolve("/app/echo"));
+        session.getBasicRemote().sendText("trigger onMessage -> onError -> onClose");
+        session.close();
+        assertTrue(clientSocket.closeLatch.await(5, TimeUnit.SECONDS));
+
+        ClassLoader webAppClassLoader = webapp.getClassLoader();
+        assertThat(event, getClassLoader(event), is(webAppClassLoader));
```

Comment:
You have the `event`, you have `clientSocket`, you only need to do `clientSocket.map.get(event).get()`.
No reflection, super-simple.

Reply:
The map is coming from the server side so it won't be in the clientSocket.

Reply:
No that doesn't work, because they are created with different classloaders they both have different instances of the static.